### PR TITLE
Added multithreading support for notebook testing and switched to using papermill, among other changes.

### PR DESCRIPTION
**Pull request by ivanmkc**
1. Switch to using papermill to run notebooks instead of nbconvert. Apparently, the reason way no output is printed during notebook execution is that nbconvert doesn't support it. See https://github.com/jupyter/nbconvert/issues/648
2. Used ThreadPoolExecutor to execute notebooks in parallel, vastly reducing runtime for multiple notebooks. The side-effect of this is that logging in real-time won't make sense as all the results will be interleaved.
3. Create separate environments for each notebook to reduce dependency conflict and enable independent testing.
4. Fixed issues with a few notebooks that had a fixed kernel embedded in its JSON.
5. Replaced pip with pip3 as pip3 doesn't work with the dynamically generated Python environments. It is unclear to me why this is. May be related to nteract/papermill#625
6. Also, the ai-platform-unified/notebooks/official/ml_metadata/sdk-metric-parameter-tracking-for-locally-trained-models.ipynb notebook was failing when deps where installed in one line. Basically, the dep resolver was taking forever. Separating the installs helped. Why? I'm not sure.
7. Fixed notebook linting Github action by changing fetch-depth.
8. Reformatted results to look like this:
```
=== RESULTS ===

file                                                              status    duration    error
----------------------------------------------------------------  --------  ----------  -------
notebook_template.ipynb                                           PASSED    00:05:11    --
automl-tabular-classification.ipynb                               PASSED    01:21:15    --
automl-text-classification.ipynb                                  PASSED    04:20:15    --
custom-tabular-bq-managed-dataset.ipynb                           PASSED    00:29:31    --
sdk-custom-image-classification-batch.ipynb                       PASSED    00:35:16    --
sdk-custom-image-classification-online.ipynb                      PASSED    00:25:58    --
gapic-custom_image_classification_batch_explain.ipynb             PASSED    00:29:34    --
gapic-custom_image_classification_online_explain.ipynb            PASSED    00:23:14    --
gapic-custom_tabular_regression_batch_explain.ipynb               PASSED    00:22:39    --
gapic-custom_tabular_regression_online_explain.ipynb              PASSED    00:11:07    --
gapic-feature-store.ipynb                                         PASSED    00:12:41    --
sdk-metric-parameter-tracking-for-custom-jobs.ipynb               PASSED    00:21:16    --
sdk-metric-parameter-tracking-for-locally-trained-models.ipynb    PASSED    00:00:44    --
model_monitoring.ipynb                                            PASSED    00:15:17    --
automl_tabular_classification_beans.ipynb                         PASSED    00:01:05    --
control_flow_kfp.ipynb                                            PASSED    00:00:28    --
google-cloud-pipeline-components_automl_images.ipynb              PASSED    00:00:25    --
google-cloud-pipeline-components_automl_tabular.ipynb             PASSED    00:00:22    --
google-cloud-pipeline-components_automl_text.ipynb                PASSED    00:00:49    --
google_cloud_pipeline_components_model_train_upload_deploy.ipynb  PASSED    00:00:59    --
lightweight_functions_component_io_kfp.ipynb                      PASSED    00:00:49    --
metrics_viz_run_compare_kfp.ipynb                                 PASSED    00:01:53    --
pipelines_intro_kfp.ipynb                                         PASSED    00:01:20    --
gapic-vizier-multi-objective-optimization.ipynb                   PASSED    00:00:29    --

=== END RESULTS===
```

### CODE REVIEW

**Comment by ivanmkc** [comment]
Pending resolution of this bug: https://github.com/nteract/papermill/issues/625

File: .cloud-build/ExecuteChangedNotebooks.py [hunk 1]
```diff
@@ -84,41 +180,68 @@ def run_changed_notebooks(
...
+            )
+            with concurrent.futures.ThreadPoolExecutor(max_workers=None) as executor:
+                notebook_execution_results = list(
+                    executor.map(
+                        functools.partial(
+                            execute_notebook,
+                            artifacts_path,
+                            variable_project_id,
+                            variable_region,
+                            False,
```
Comment (ivanmkc) [inline .cloud-build/ExecuteChangedNotebooks.py:200]: Don't show logs when running notebooks in parallel due to interleaved logs from multiple notebooks.

File: .cloud-build/ExecuteChangedNotebooks.py [hunk 2]
```diff
@@ -84,41 +180,68 @@ def run_changed_notebooks(
...
+    print("\n=== RESULTS ===\n")
+
+    notebooks_sorted = sorted(
+        notebook_execution_results,
+        key=lambda result: result.is_pass,
+        reverse=True,
+    )
+    # Print results
+    print(
+        tabulate(
```
Comment (ivanmkc) [inline .cloud-build/ExecuteChangedNotebooks.py:230]: Switched to using `tabulate` to nicely format notebooks.

File: .cloud-build/ExecuteChangedNotebooks.py [hunk 3]
```diff
@@ -13,20 +13,112 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import argparse
+import dataclasses
```
Comment (tswast) [inline .cloud-build/ExecuteChangedNotebooks.py:16]: GitHub won't let me put the comment on the line, but the first line `#!/bin/bash` is throwing off GitHub's syntax highlighting.

Should be `#!/usr/bin/env python`

File: .cloud-build/ExecuteNotebook.py [hunk 4]
```diff
@@ -1,20 +1,100 @@
+import json
```
Comment (tswast) [inline .cloud-build/ExecuteNotebook.py:1]: You're missing a license header.

File: .cloud-build/ExecuteChangedNotebooks.py (hunk 3 above)
Comment (ivanmkc) [reply .cloud-build/ExecuteChangedNotebooks.py:16]: Good catch, let me fix those.

File: .cloud-build/ExecuteNotebook.py (hunk 4 above)
Comment (ivanmkc) [reply .cloud-build/ExecuteNotebook.py:1]: Thanks, will add.

File: .cloud-build/notebook-execution-test-cloudbuild.yaml [hunk 5]
```diff
@@ -30,7 +30,7 @@ steps:
     entrypoint: /bin/sh
     args: 
     - -c
-    - 'if [ $(ls -pR "/workspace/${BUILD_ID}"  | grep -v / | grep -v ^$ | wc -l) -ne 0 ]; then gsutil rsync -r "/workspace/${BUILD_ID}" "gs://${_GCS_ARTIFACTS_BUCKET}/test-artifacts/PR_${_PR_NUMBER}/BUILD_${BUILD_ID}/"; else echo "No artifacts to copy."; fi'
+    - 'if [ $(ls -pR "/workspace/${BUILD_ID}"  | grep -v / | grep -v ^$ | wc -l) -ne 0 ]; then gsutil -m -q rsync -r "/workspace/${BUILD_ID}" "gs://${_GCS_ARTIFACTS_BUCKET}/test-artifacts/PR_${_PR_NUMBER}/BUILD_${BUILD_ID}/"; else echo "No artifacts to copy."; fi'
```
Comment (ivanmkc) [inline .cloud-build/notebook-execution-test-cloudbuild.yaml:33]: Improve speed and make it quiet

**Pull request merged**